### PR TITLE
Core Interfaces

### PR DESCRIPTION
**Pull request by nullorvoid**
# Description

Added the core interfaces used by the tile system, these are mostly complete, they need a review to make sure I haven't missed any.  The only two things that I have changed between these and the original design is that I have added ITile, IArea, and ILevel which will be in addition to the base classes that will be implemented later; and I have changed the signature of the factory to be a little more user friendly.
# What Changed
- Added Base, Creational, Entity Management, Solvers, and Tile Change interfaces
- Added EventArgs to go with them
- Created the first Util class, which is the Tile Change Type

### CODE REVIEW

File: TileSystem/Interfaces/Creation/ICreateTiles.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	/// <summary>
+	/// ICreateTiles is used by the level to create tiles
+	/// using a normal creational pattern
+	/// 
+	/// The area, x, y will be where the tile is created
+	/// </summary>
+	public interface ICreateTiles
+	{
+		event EventHandler<TileCreatedArgs> TileCreated;
+		ITile CreateTile(IArea area, int x, int y, string type, string variation, params object[] properties);
```
Comment (kefniark) [inline TileSystem/Interfaces/Creation/ICreateTiles.cs]: it seems to be the only interface that use directly **[x, y]**

Is it possible to create an Interface **ILocation** to abtract that and not enforce this 2d array representation (the last week discussion with @christopherobin ^^)
Allowing nodes, 3d or whatever other crazy idea and let that to the implementation

File: TileSystem/Interfaces/Management/IManageEntities.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+
+namespace TileSystem.Interfaces.Management
+{
+	/// <summary>
+	/// Used by the tiles to manage the entities being moved
+	/// between them
+	/// </summary>
+	public interface IManageEntities
```
Comment (kefniark) [inline TileSystem/Interfaces/Management/IManageEntities.cs:11]: This should probably be merged with **ICreateEntities / ICreateTiles**

**Comment by kefniark** [comment]
With should let IPosition up to the implementation.
The abstract implementation 2d, 3d, nodes, should be separated in another namespace

**Comment by kefniark** [comment]
We should write documentation with graphics to explicit this structure :

> Level (board) -> Area (floor) -> Tiles -> Entities

With basic explanation:
- an area has other neighbours area (nodes)
- a tile has other neighbours tiles (nodes)
- an entity can move across tiles and areas
- a tile can contains 0-n entities

File: TileSystem/Interfaces/Creation/ICreateTiles.cs (hunk 1 above)
Comment (nullorvoid) [inline TileSystem/Interfaces/Creation/ICreateTiles.cs]: Fixed, now we have a simple interface that is IPosition, which you should implement comparable so we can tell if positionA == positionB

File: TileSystem/Interfaces/Management/IManageEntities.cs (hunk 2 above)
Comment (nullorvoid) [inline TileSystem/Interfaces/Management/IManageEntities.cs:11]: Will keep these separate so we can have creation and management separate for all objects in the system.

**Comment by nullorvoid** [comment]
Yes I think the best time to do this is when we implement the 2D Base Classes, it will give us nice VS diagrams to help explain this.  I also want to add documentation like the [AC docs](https://aeriscloud.github.io/AerisCloud/)

File: TileSystem/Interfaces/Creation/IAreaFactory.cs [hunk 3]
```diff
@@ -3,7 +3,8 @@
 namespace TileSystem.Interfaces.Creation
 {
 	/// <summary>
-	/// 
+	/// Factory for areae creation, defined by the framework to be easily linked
```
Comment (kefniark) [inline TileSystem/Interfaces/Creation/IAreaFactory.cs]: stop writing in latin please `areae`

Comment (nullorvoid) [inline TileSystem/Interfaces/Creation/IAreaFactory.cs]: ^^

**Comment by kefniark** [comment]
[LGTM](http://lgtm.in/i/gSpmDbc8q)

**Comment by kefniark** [comment]
I let you fix stuff and work more this weekend o7

**Pull request merged**